File: onebusaway-android/src/main/java/org/onebusaway/android/ui/survey/SurveyPreferences.java
package org.onebusaway.android.ui.survey;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.Date;
import java.util.UUID;

/**
 * Utility class for managing survey-related preferences using SharedPreferences.
 * This class handles storing and retrieving user UUIDs and survey reminder dates.
 */
public class SurveyPreferences {

    private static final String PREFS_NAME = "survey_pref";
    private static final String UUID_KEY = "my_uuid";
    private static final String SURVEY_REMINDER_DATE_KEY = "survey_reminder_day";


    /**
     * Saves the user's UUID to the shared preferences.
     *
     * @param uuid the UUID to be saved in shared preferences
     */
    private static void saveUserUUIDHelper(Context context, UUID uuid) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(UUID_KEY, uuid.toString());
        editor.apply();
    }

    /**
     * Retrieves the user's UUID from shared preferences.
     *
     * @return the saved UUID as a string, or null if not found
     */
    private static String getUserUUIDHelper(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String uuidString = prefs.getString(UUID_KEY, null);
        return uuidString != null ? UUID.fromString(uuidString).toString() : null;
    }

    /**
     * Saves the given survey reminder date in SharedPreferences.
     *
     * @param date The date to be saved as a reminder
     */
    public static void setSurveyReminderDate(Context context,Date date) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putLong(SURVEY_REMINDER_DATE_KEY, date.getTime());
        editor.apply();
    }

    /**
     * Retrieves the survey reminder date from SharedPreferences.
     *
     * @return The stored reminder date as a long (milliseconds since epoch), or -1 if not set
     */
    public static Long getSurveyReminderDate(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        return prefs.getLong(SURVEY_REMINDER_DATE_KEY, -1);
    }


    /**
     * Retrieves the user's UUID. If the UUID is not already stored, generates a new UUID and saves it.
     *
     * @return The user's UUID as a String.
     */
    public static String getUserUUID(Context context) {
        if (SurveyPreferences.getUserUUIDHelper(context) == null) {
            UUID uuid = UUID.randomUUID();
            SurveyPreferences.saveUserUUIDHelper(context, uuid);
        }
        return SurveyPreferences.getUserUUIDHelper(context);
    }
}

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/survey/SurveyPreferences.kt
/*
 * Copyright (C) 2024-2026 Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.ui.survey

import android.content.Context
import java.util.Date
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import org.onebusaway.android.preferences.PreferencesRepository

/**
 * Survey-related preferences (the user's survey UUID and the reminder date). Backed by the
 * DataStore [PreferencesRepository] (storage-modernization) rather than the old separate `survey_pref`
 * SharedPreferences file; values from that file are migrated once on first access so existing users
 * keep their survey identity.
 *
 * Injectable (#1636): the [PreferencesRepository] is constructor-injected rather than resolved via a
 * static `Application.get()` graph handle. Consumers obtain it through Hilt (e.g. [SurveyViewModel])
 * and pass it into the pure `SurveyUtils` gating helpers.
 */
@Singleton
class SurveyPreferences @Inject constructor(
    private val prefs: PreferencesRepository,
) {

    /**
     * Saves the given survey reminder date.
     *
     * @param date The date to be saved as a reminder
     */
    fun setSurveyReminderDate(date: Date) {
        prefs.setLong(SURVEY_REMINDER_DATE_KEY, date.time)
    }

    /**
     * Retrieves the survey reminder date.
     *
     * @return The stored reminder date as a long (milliseconds since epoch), or -1 if not set
     */
    fun getSurveyReminderDate(context: Context): Long {
        migrateLegacyIfNeeded(context)
        return prefs.getLong(SURVEY_REMINDER_DATE_KEY, -1)
    }

    /**
     * Retrieves the user's UUID, generating and storing one if absent.
     *
     * @return The user's UUID as a String.
     */
    fun getUserUUID(context: Context): String {
        migrateLegacyIfNeeded(context)
        prefs.getString(UUID_KEY, null)?.let { return it }
        return UUID.randomUUID().toString().also { prefs.setString(UUID_KEY, it) }
    }

    /** One-time copy of the old separate `survey_pref` SharedPreferences values into DataStore. */
    private fun migrateLegacyIfNeeded(context: Context) {
        if (prefs.getBoolean(LEGACY_MIGRATED_KEY, false)) return
        val legacy = context.getSharedPreferences(LEGACY_PREFS_NAME, Context.MODE_PRIVATE)
        legacy.getString(UUID_KEY, null)?.let { prefs.setString(UUID_KEY, it) }
        legacy.getLong(SURVEY_REMINDER_DATE_KEY, -1)
            .takeIf { it != -1L }
            ?.let { prefs.setLong(SURVEY_REMINDER_DATE_KEY, it) }
        prefs.setBoolean(LEGACY_MIGRATED_KEY, true)
    }

    companion object {
        private const val LEGACY_PREFS_NAME = "survey_pref"
        private const val UUID_KEY = "my_uuid"
        private const val SURVEY_REMINDER_DATE_KEY = "survey_reminder_day"
        private const val LEGACY_MIGRATED_KEY = "survey_pref_migrated"
    }
}
